parser-typechecker/src/Unison/PrettyPrintEnv.hs

```diff
@@ -34,10 +36,43 @@ instance Show PrettyPrintEnv where
   show _ = "PrettyPrintEnv"
 
 fromNames :: Int -> Names -> PrettyPrintEnv
-fromNames length names = PrettyPrintEnv terms' types' where
-  terms' r = safeHead . Set.map HQ'.toHQ $ (Names.termName length r names)
-  types' r = safeHead . Set.map HQ'.toHQ $ (Names.typeName length r names)
-
+fromNames len names = PrettyPrintEnv terms' types' where
+  terms' r = shortestName . Set.map HQ'.toHQ $ (Names.termName len r names)
+  types' r = shortestName . Set.map HQ'.toHQ $ (Names.typeName len r names)
+  shortestName ns = safeHead $ sortOn (length . HQ.toString) (toList ns)
+
+fromSuffixNames :: Int -> Names -> PrettyPrintEnv
+fromSuffixNames len names = fromNames len (Names.suffixify names)
+
+fromNamesDecl :: Int -> Names -> PrettyPrintEnvDecl
+fromNamesDecl len names = 
+  PrettyPrintEnvDecl (fromNames len names) (fromSuffixNames len names)
+
+-- A pair of PrettyPrintEnvs:
+--   - suffixifiedPPE uses the shortest unique suffix
+--   - unsuffixifiedPPE uses the shortest full name
+data PrettyPrintEnvDecl = PrettyPrintEnvDecl {
+  unsuffixifiedPPE :: PrettyPrintEnv,
+  suffixifiedPPE :: PrettyPrintEnv
+  } deriving Show
+
+-- declarationPPE uses the full name for references that are
+-- part the same cycle as the input reference, used to ensures
+-- recursive definitions are printed properly, for instance:
+--
+-- foo.bar x = foo.bar x
+-- and not
+-- foo.bar x = bar x
```

Comment:
— the full, probably-relative name, right?  Are these always relative names btw?